029-across-step/proposal.md

```diff
@@ -0,0 +1,321 @@
+full matrix completes.
+
+`attempts` binds to the inner step because it doesn't seem to make a whole lot
+of sense to retry the *entire matrix* because one step failed. The individual
+steps should be retried instead.
+
+`timeout` binds to the inner step because otherwise a `max_in_flight` could
+cause the timeout to be exceeded before some steps even get a chance to run.
+
+```yaml
+task: unit
+timeout: 1h # interrupt the task after 1 hour
+attempts: 3 # attempt the task 3 times
+across:
+- var: go_version
+  values: [1.12, 1.13]
+on_failure: # do something after all steps complete and at least one failed
+```
+
+To apply `ensure` and `on_*` hooks to the nested step, rather than the `across`
+step modifier, the `do:` step may be utilized:
+
+```yaml
+do:
+- task: unit
+  on_failure: # runs after each individual step completes and fails
+across:
+- var: go_version
+  values: [1.12, 1.13]
+on_failure: # runs after all steps complete and at least one failed
+```
+
+This can be rewritten in a slightly more readable syntax by placing the `do:`
+below the `across:`:
+
+```yaml
+across:
+- var: go_version
+  values: [1.12, 1.13]
+do:
+- task: unit
+  on_failure: # runs after each individual step completes and fails
+on_failure: # runs after all steps complete and at least one failed
+```
+
+### Failing fast
+
+When a step within the matrix fails, the `across` step will continue to run the
+remaining steps in the matrix. However, the `across` step itself will still
+fail after all steps have completed.
+
+With `fail_fast: true` applied to the `across` step, execution will halt on the
+first failure, and any currently running steps (via `max_in_flight`) will be
+interrupted:
+
+```yaml
+task: unit
+across:
+- var: go_version
+  values: [1.12, 1.13]
+fail_fast: true
+```
+
+Note: this is the first time a step *modifier* has had additional sibling
+fields. In the event of a field conflict, the `do:` step may be utilized as a
+work-around.
+
+### Var scoping and shadowing
+
+The inner step will be run with a local var scope that inherits from the outer
+scope and is initialized with the across step's var values.
+
+Given that it runs with its own scope, it follows that the vars in the local
+scope, along with any vars newly set within the scope, are not accessible
+outside of the `across` step.
+
+When a var set by the `across` step shadows outer vars, a warning will be
+printed.
+
+Example:
+
+```yaml
+plan:
+- get_var: foo
+  trigger: true
+- run: print
+  type: debug
+  params: {value: ((.:foo))}
+- across:
+  - var: foo
+    values: [one, two]
+  do:
+  - run: print
+    type: debug
+    params: {value: ((.:foo))}
+- run: print
+  type: debug
+  params: {value: ((.:foo))}
+```
+
+Assuming the `get_var` step produces a value of `zero`, this build plan should
+result in the following output:
+
+```
+zero
+WARNING: across step shadows local var 'foo'
+one
+two
+zero
+```
+
+
+## Open Questions
+
+* n/a
```

Comment:
Something that didn't really come up in the proposal is how we handle steps that emit artifacts (e.g. `get`, `task.outputs`, prototype outputs). Currently, the implementation of the static `across` step runs each iteration in its own local artifact scope. This means that any artifacts go out of scope past the `across` step and cannot be used.

It'd be cool if there was a way to "fan-in" on artifacts emitted within an `across` step. For instance, suppose there was a `prototype` (or just a `task`) for compiling a Go binary that outputs an artifact called `binary`. You might want to compile it for multiple platforms/architectures, which could make use of the `across` step:

```yaml
across:
- var: platform
  values: [linux, darwin, windows]
- var: arch
  values: [amd64, arm64]
run: compile
type: go
params:
  package: repo/cmd/my-program
  platform: ((.:platform))
  arch: ((.:arch))
outputs: [binary]
```

(side-note: it might make more sense from a performance PoV for a `go` prototype to support compiling multiple platforms/architectures in a single `run` step where possible, but bear with me for the example)

The trouble is - under the current implementation, there's no way to access `binary` outside of the `across` step. Even if we used a shared artifact scope for each iteration, it'd just be "last write wins" (so, with no parallelism, `binary` would be the result for `(windows, arm64)` only) - since they all share a name, the results would clobber each other. This could be avoided by encoding more information in the output name via an `output_mapping` (e.g. name the artifacts `binary-linux-amd64`, `binary-linux-arm64`, ...) - but that only works for simple values like strings/numbers.

So, what if we let artifacts in an `across` step share a name, but be uniquely identified by the set of vars that were used to produce that artifact (in the same way that instanced pipelines share a name and are uniquely identified by their `instance_vars`). A possible syntax for referencing the binary created for `(linux, amd64)` could be `binary{platform: linux, arch: amd64}`

Here, `binary` really refers to a matrix of outputs (that mirrors the `across` matrix). What if we also provided a way to filter down that matrix to get at a subset of the outputs. e.g. `binary{platform: windows}` would give the (2) outputs built for windows, and `binary{arch: arm64}` would give the (3) outputs built for arm64.

So, suppose you wanted to upload a github release containing *all* 6 binaries produced in the matrix. You could do something like:

```yaml
put: release
params:
  globs: [binary/*] # or maybe [binary{*}/*] to make it clear it's a matrix
```

Suppose you wanted to bundle the different architectures separately - then, you could do:

```yaml
across:
- var: arch
  values: [amd64, arm64]
put: ((.:arch))-bundle
params:
  globs: [binary{arch: ((.:arch))}/*]
```

This may warrant a separate proposal but figured I'd bring it up here since it's definitely a limitation of the `across` step as it stands

Reply:
One idea, maybe we could re-use the concept of `output_mappings`. Similar to how a task defined in a file can have its outputs mapped by an `output_mappings` configuration in the pipeline referencing that task, an `across` step could have its outputs made available via a similar mapping.

As an example:
```yaml
across:
- var: arch
  values: [amd64, arm64]
- var: version
  values: [1.0, 1.1]

task: create-artifact
output_mapping:
  output: task-artifact

across_output_mapping:
- vars: # match vars from an execution of the task
    arch: amd64
    version: 1.0
  outputs: # map an output from the modified step to an output of the across step
    task-artifact: amd64-bundle-v1.0
```

This would be a fairly verbose if you wanted to map every output from the matrix of steps:
```yaml
# [...]
across_output_mapping:
- vars: {arch: amd64, version: 1.0}
  outputs: {task-artifact: amd64-bundle-v1.0}
- vars: {arch: arm64, version: 1.0}
  outputs: {task-artifact: arm64-bundle-v1.0}
- vars: {arch: amd64, version: 1.1}
  outputs: {task-artifact: amd64-bundle-v1.1}
- vars: {arch: arm64, version: 1.1}
  outputs: {task-artifact: arm64-bundle-v1.1}
```

Though it would possibly simpler to implement, without relying on prototypes or other Concourse functionality.

When, or if, [this limitation](https://concourse-ci.org/across-step.html#limitations) on using vars in `get:` and `put:` steps (eg. `get: bundle-((.:arch))-((.:version))`) is lifted, the `across_output_mapping` could be updated to allow for the use of vars as well, simplifying the configuration to something like...

```yaml
# [...]
across_output_mapping:
- outputs:
    task-artifact: ((.:arch))-bundle-v((.:version))
```